File: scripts/foo.ts
import { safetensors, tokenizers, WeightMapper } from "@jax-js/loaders";
import { defaultDevice, init, numpy as np, tree } from "@jax-js/jax";
import { readFileSync } from "node:fs";
import {
	emptyKVCache,
	fromSafetensors,
	MAX_CONTEXT_LEN,
	runAttention,
	type Gemma3,
	type Gemma3Attention,
} from "$lib/gemma3/model";
import { runInference } from "$lib/gemma3/inference";
import { AutoTokenizer, PreTrainedTokenizer } from "@huggingface/transformers";

/* eslint-disable @typescript-eslint/no-unused-vars */

function test_fromFile() {
	const weightMapper = new WeightMapper({
		autoCamelCase: true,
	});

	const buf = readFileSync("weights/model.f16.safetensors");
	const file = safetensors.parse(buf); // => { tensors: { ... } };
	const mappedWeights = weightMapper.mapObject(file.tensors);
	const weights = safetensors.toNested(mappedWeights);

	console.log(Object.keys(weights.model.layers[0]));

	for (const key of Object.keys(weights.model.layers[0])) {
		console.log(Object.keys(weights.model.layers[0][key]));
	}
}

function test_fromSafetensors() {
	const buf = readFileSync("weights/model.f16.safetensors");
	const file = safetensors.parse(buf);
	const weights = fromSafetensors(file);

	console.log(weights.layers[0]);
}

function test_jaxjs_tokenizer() {
	const data = readFileSync("tokenizer/tokenizer.model");
	const tokenizer = tokenizers.Unigram.fromBinary(data);

	console.log(tokenizer);

	const text = "Plants create energy through a process known as";
	console.log("Tokens: ", tokenizer.encode(text));
}

async function test_transformers_tokenizer() {
	const tokenizer = await PreTrainedTokenizer.from_pretrained("./tokenizer/");

	const text = "Plants create energy through a process known as";
	console.log("Tokens: ", tokenizer.encode(text));
}

function test_dot() {
	const N = 10;
	const q = np.ones([N, 4, 32]);
	const k = np.ones([N, 32]);

	const q_kT = np.dot(q, k.transpose());
	console.log(q_kT.shape);
}

function test_mask() {
	const offset = 1;
	const S = 2;
	const N = 4;
	const mask = np
		.arange(N)
		.sub((offset + S) % N)
		.add(N)
		.mod(N)
		.add(offset + S - N);
	// console.log(mask.js());
	const sequencePositions = np.arange(offset, offset + S).reshape([S, 1]);
	const slotPositions = np
		.arange(N)
		.sub((offset + S) % N)
		.add(N)
		.mod(N)
		.add(offset + S - N); // Cache slots mapped to their actual positions
	console.log(sequencePositions.ref.js());
	console.log(slotPositions.ref.js());
	console.log(
		slotPositions.ref.lessEqual(sequencePositions).mul(slotPositions.greaterEqual(0)).js(),
	);
}

function test_rotate_half() {
	// console.log(rotateHalf(np.arange(6)).js());
}

function test_attention() {
	const weights: Gemma3Attention = {
		qProj: { weight: np.ones([1024, 640]) },
		kProj: { weight: np.ones([256, 640]) },
		vProj: { weight: np.ones([256, 640]) },
		oProj: { weight: np.ones([640, 1024]) },
		qNorm: { gamma: np.ones([256]) },
		kNorm: { gamma: np.ones([256]) },
	};
	const isSlidingAttention = true;
	const kvCache = emptyKVCache(isSlidingAttention ? 512 : MAX_CONTEXT_LEN, 256);
	const x = np.ones([999, 640]);
	console.log(runAttention(weights, kvCache, x, 0, isSlidingAttention)[0].shape);
}

// Don't use, way too slow!
async function test_local_inference() {
	const devices = await init("cpu");
	defaultDevice("cpu");

	console.log("Running...");

	const data = readFileSync("weights/model.f16.safetensors");
	const file = safetensors.parse(data);
	const weights: Gemma3 = fromSafetensors(file);
	const text = "Plants create energy through a process known as";

	console.log("Loaded weights");

	const tokenizer = await AutoTokenizer.from_pretrained("./tokenizer/");

	await runInference(weights, tokenizer, text);

	tree.dispose(weights);
}

function test_modules() {
	const buf = readFileSync("weights/model.f16.safetensors");
	const file = safetensors.parse(buf); // => { tensors: { ... } };
	const weights = fromSafetensors(file);

	console.log(weights.layers[0].inputLayernorm.gamma.js());
}

function test_reshape() {
	const positions = np.arange(10).reshape([-1, 1]);
	// console.log(positions.js());

	// const x = np.arange(10).reshape([1, 10]);
	// const y = np.tile(x, [3, 1]);
	// console.log(y.js());

	const x = np.arange(10).sub(np.arange(5).reshape([5, 1]));
	console.log(x.js());
}

test_mask();

File: scripts/foo.py
import types
from typing import Any

import torch
from transformers import AutoModelForCausalLM, AutoTokenizer
from safetensors import safe_open

from scripts.reference import get_reference_f16


def test_weights():
    with (
        safe_open("weights/model.bf16.safetensors", framework="pt") as b,
        safe_open("weights/model.f16.safetensors", framework="pt") as f,
    ):
        # Get a specific tensor
        for key in b.keys():
            b_tensor = b.get_tensor(key).to(torch.float32)
            _f_tensor = f.get_tensor(key).to(torch.float32)

            mask = torch.abs(b_tensor) < 1e-5
            indices = torch.where(mask)
            if len(indices[0]) > 0:
                print(f"{key}: FAILED")
                print(indices)
                print(b_tensor[mask])

            # if not torch.allclose(b_tensor, f_tensor):
            #     print(f"{key}: FAILED")
            #     mask = ~torch.isclose(b_tensor, f_tensor)
            #     indices = torch.where(mask)

            #     print(indices)  # Outputs the index/indices
            #     print(b_tensor[mask], f_tensor[mask])  # Shows the violating values


def test_dtypes():
    devices = ["cpu"]
    if torch.cuda.is_available():
        devices.append("cuda")
    if torch.backends.mps.is_available():
        devices.append("mps")

    dtypes = [torch.float32, torch.float16, torch.bfloat16]

    for device in devices:
        for dtype in dtypes:
            try:
                x = torch.ones(4, 4, dtype=dtype, device=device)
                _y = x @ x  # matmul is a good stress test
                print(f"{device} + {dtype}: OK")
            except Exception as e:
                print(f"{device} + {dtype}: FAILED — {e}")


def test_overunderflow():
    tokenizer_path = "./tokenizer/"
    model_path = "./model_f16/"

    tokenizer = AutoTokenizer.from_pretrained(tokenizer_path)
    model = AutoModelForCausalLM.from_pretrained(
        model_path,
        # dtype=torch.float32,
        device_map="auto",
        # attn_implementation="eager",
    )

    # After conversion, check the weight distribution
    for name, param in model.named_parameters():
        if param.isinf().any():
            print(f"{name}: has inf (overflow during cast)")
        if param.isnan().any():
            print(f"{name}: has nan")
        # print(f"{name}: max={param.abs().max():.3f}")

    import inspect

    for name, module in model.named_modules():
        if "layers.7" not in name:
            continue

        if name == "model.layers.7":
            print(inspect.getsource(module.__class__.forward))

        def hook(m, input, output, n=name):
            t = output if isinstance(output, torch.Tensor) else output[0]
            if len(input) > 0 and isinstance(input[0], torch.Tensor):
                print(input[0].abs().max())
            print(
                f"{n}: max={t.abs().max():.1f}, has_inf={t.isinf().any()}, has_nan={t.isnan().any()}"
            )

        module.register_forward_hook(hook)

        # def hook(m, input, output, n=name):
        #     # if "layers.7" not in n:
        #     #     return
        #     t = output if isinstance(output, torch.Tensor) else output[0]
        #     if torch.isnan(t).any():
        #         print(f"{n}: has nan")
        #     if torch.isinf(t).any():
        #         print(f"{n}: has inf")
        #     if t.abs().max() < 1e-6:
        #         print(f"{n}: suspiciously small max={t.abs().max()}")

        # module.register_forward_hook(hook)

    text = "Plants create energy through a process known as"
    input_ids = tokenizer(text, return_tensors="pt").to(model.device)

    output = model.generate(
        **input_ids,
        cache_implementation="static",
        max_length=(input_ids["input_ids"].shape[1] + 10),
    )
    print(tokenizer.decode(output[0], skip_special_tokens=True))


def test_tokenizer():
    tokenizer = AutoTokenizer.from_pretrained("./tokenizer/")

    text = "Plants create energy through a process known as"
    print("Tokens: ", tokenizer(text))


def test_reference():
    def rotate_half(x):
        """Rotates half the hidden dims of the input."""
        x1 = x[..., : x.shape[-1] // 2]
        x2 = x[..., x.shape[-1] // 2 :]
        return torch.cat((-x2, x1), dim=-1)

    def apply_rotary_pos_emb(q, k, cos, sin, unsqueeze_dim=1):
        """Applies Rotary Position Embedding to the query and key tensors.

        Args:
            q (`torch.Tensor`): The query tensor.
            k (`torch.Tensor`): The key tensor.
            cos (`torch.Tensor`): The cosine part of the rotary embedding.
            sin (`torch.Tensor`): The sine part of the rotary embedding.
            unsqueeze_dim (`int`, *optional*, defaults to 1):
                The 'unsqueeze_dim' argument specifies the dimension along which to unsqueeze cos[position_ids] and
                sin[position_ids] so that they can be properly broadcasted to the dimensions of q and k. For example, note
                that cos[position_ids] and sin[position_ids] have the shape [batch_size, seq_len, head_dim]. Then, if q and
                k have the shape [batch_size, heads, seq_len, head_dim], then setting unsqueeze_dim=1 makes
                cos[position_ids] and sin[position_ids] broadcastable to the shapes of q and k. Similarly, if q and k have
                the shape [batch_size, seq_len, heads, head_dim], then set unsqueeze_dim=2.
        Returns:
            `tuple(torch.Tensor)` comprising of the query and key tensors rotated using the Rotary Position Embedding.
        """
        cos = cos.unsqueeze(unsqueeze_dim)
        sin = sin.unsqueeze(unsqueeze_dim)
        q_embed = (q * cos) + (rotate_half(q) * sin)
        k_embed = (k * cos) + (rotate_half(k) * sin)
        return q_embed, k_embed

    def make_patched_forward(original_forward):
        def patched_forward(
            self,
            hidden_states: torch.Tensor,
            position_embeddings: torch.Tensor = None,
            attention_mask: torch.Tensor | None = None,
            past_key_values: Any | None = None,
            **kwargs: Any,
        ) -> tuple[torch.Tensor, torch.Tensor | None, tuple[torch.Tensor] | None]:
            input_shape = hidden_states.shape[:-1]
            hidden_shape = (*input_shape, -1, self.head_dim)

            query_states = self.q_proj(hidden_states).view(hidden_shape).transpose(1, 2)
            key_states = self.k_proj(hidden_states).view(hidden_shape).transpose(1, 2)
            value_states = self.v_proj(hidden_states).view(hidden_shape).transpose(1, 2)

            query_states = self.q_norm(query_states)
            key_states = self.k_norm(key_states)

            cos, sin = position_embeddings
            print("cos")
            print(cos)
            print("sin")
            print(sin)
            query_states, key_states = apply_rotary_pos_emb(
                query_states, key_states, cos, sin
            )
            print("query_states")
            print(query_states)
            print("key_states")
            print(key_states)

            return original_forward(
                # self,
                hidden_states,
                position_embeddings,
                attention_mask,
                past_key_values,
                **kwargs,
            )

        return patched_forward

    tokenizer, model = get_reference_f16()

    layer = model.model.layers[0].self_attn
    layer.forward = types.MethodType(make_patched_forward(layer.forward), layer)

    for name, module in model.named_modules():
        if "layers.0" not in name:
            continue

        def hook(m, input, output, n=name):
            i = input[0] if len(input) > 0 else None
            o = output if isinstance(output, torch.Tensor) else output[0]
            print(n)
            print(i)
            print(o)

        module.register_forward_hook(hook)

    # text = "Plants create energy through a process known as"
    text = "The capital of France is "
    input_ids = tokenizer(text, return_tensors="pt").to(model.device)

    print(input_ids["input_ids"])

    output = model.generate(
        **input_ids,
        cache_implementation="static",
        max_length=(input_ids["input_ids"].shape[1] + 1),
    )
    print(tokenizer.decode(output[0], skip_special_tokens=False))


def test_modules():
    tokenizer, model = get_reference_f16()

    import inspect

    print(
        inspect.getsource(
            model.get_submodule("model.layers.0.self_attn").__class__.forward
        )
    )
    # input_layernorm_0 = model.get_submodule("model.layers.0.input_layernorm")
    # print(input_layernorm_0.state_dict())


if __name__ == "__main__":
    test_reference()
